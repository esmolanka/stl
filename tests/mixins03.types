
type SimpleResource a = mixin
  | FromFile   : String
  | FromMemory : Int
  | HardCoded  : a

type DBResource = mixin
  | FromDB : {database : String, table : String}

type EtherResource =
  mixin < FromEther : exists a. a >

type Resource a =
  | mix SimpleResource a
  | mix DBResource
  | mix EtherResource

#eval Resource

type Const b a = b

type ResourceLoadEnum = variant of (Const Unit)
  | mix SimpleResource Unit
  | mix DBResource
  | mix EtherResource

#eval ResourceLoadEnum
