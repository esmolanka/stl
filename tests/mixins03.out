tests/mixins03.types:18:1-18:15:
  Normalised:
    λ a. ∀ (ρ : Row).
      < FromFile? : String
      | FromMemory? : Int
      | HardCoded? : a
      | FromDB? : ∃ (ρ : Row). {database : String, table : String | ρ}
      | FromEther? : ∃ a. a
      | ... ρ
      >
  Kind:
    Type -> Type
tests/mixins03.types:27:1-27:23:
  Normalised:
    ∀ (ρ : Row).
      < FromFile? : Unit
      | FromMemory? : Unit
      | HardCoded? : Unit
      | FromDB? : Unit
      | FromEther? : Unit
      | ... ρ
      >
  Kind:
    Type